File: library-java/src/main/java/com/libraryjava/service/BookService.java
package com.libraryjava.service;

import com.libraryjava.domain.book.Book;
import com.libraryjava.domain.book.BookType;
import com.libraryjava.domain.user.User;
import com.libraryjava.domain.user.UserStatus;
import com.libraryjava.domain.user.loanhistory.UserLoanStatus;
import com.libraryjava.dto.book.BookLoanDto;
import com.libraryjava.dto.book.BookMakeDto;
import com.libraryjava.dto.book.BookReturnDto;
import com.libraryjava.dto.book.response.BookStatResponse;
import com.libraryjava.repository.BookRepository;
import com.libraryjava.repository.UserLoanHistoryRepository;
import com.libraryjava.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static java.util.stream.Collectors.groupingBy;

@Service
@RequiredArgsConstructor
@Transactional
public class BookService {
    private final BookRepository bookRepository;
    private final UserRepository userRepository;
    private final UserLoanHistoryRepository userLoanHistoryRepository;

    public void makeBook(BookMakeDto makeDto) {
        bookRepository.save(new Book(makeDto.name(), makeDto.type()));
    }

    public void loanBook(BookLoanDto loanDto) {
        if (userLoanHistoryRepository.findByBookNameAndUserLoanStatus(loanDto.bookName(), UserLoanStatus.LOANED).isPresent()) {
            throw new IllegalArgumentException("진작 대출되어 있는 책입니다.");
        }

        User findUser = userRepository.findByName(loanDto.userName())
                .orElseThrow(IllegalArgumentException::new);

        bookRepository.findByName(loanDto.bookName())
                .orElseThrow(IllegalArgumentException::new);

        findUser.loanBook(loanDto.bookName());
    }

    public void returnBook(BookReturnDto returnDto) {
        User findUser = userRepository.findByName(returnDto.userName())
                .orElseThrow(IllegalArgumentException::new);

        findUser.returnBook(returnDto.bookName());
    }

    @Transactional(readOnly = true)
    public int countLoanedBook() {
//        return userLoanHistoryRepository.findAllByUserLoanStatus(UserLoanStatus.LOANED).size();
        return userLoanHistoryRepository.countByUserLoanStatus(UserLoanStatus.LOANED).intValue();
    }

    @Transactional(readOnly = true)
    public List<BookStatResponse> getBookStatistics() {
//        Map<BookType, List<Book>> map = bookRepository.findAll().stream()
//                .collect(groupingBy(book -> book.getType()));
//
//        List<BookStatResponse> result = new ArrayList<>();
//        for (BookType bookType : map.keySet()) {
//            result.add(new BookStatResponse(bookType, map.get(bookType).size()));
//        }
//        return result;

//        List<BookStatResponse> responses = new ArrayList<>();
//        List<Book> books = bookRepository.findAll();
//        for (Book book : books) {
//            responses.stream()
//                    .filter(dto -> dto.getBookType() == book.getType())
//                    .findFirst()
//                    .ifPresentOrElse(BookStatResponse::plusOne, () -> responses.add(new BookStatResponse(book.getType(), 1)));
//        }
        return bookRepository.getStats();
    }
}

File: src/main/kotlin/com/library/service/book/BookService.kt
package com.library.service.book

import com.library.domain.book.Book
import com.library.domain.book.BookRepository
import com.library.domain.user.UserRepository
import com.library.domain.user.loanhistory.UserLoanHistoryRepository
import com.library.domain.user.loanhistory.UserLoanStatus
import com.library.dto.book.BookLoanRequest
import com.library.dto.book.BookRequest
import com.library.dto.book.BookReturnRequest
import com.library.dto.book.BookStatResponse
import com.library.util.fail
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class BookService(
    private val bookRepository: BookRepository,
    private val userRepository: UserRepository,
    private val userLoanHistoryRepository: UserLoanHistoryRepository,
) {
    @Transactional
    fun saveBook(request: BookRequest) {
        val newBook = Book(request.name, request.type)
        bookRepository.save(newBook)
    }

    @Transactional
    fun loanBook(request: BookLoanRequest) {
        val book = bookRepository.findByName(request.bookName) ?: fail()
        if (userLoanHistoryRepository.findByBookNameAndStatus(request.bookName, UserLoanStatus.LOANED) != null) {
            throw IllegalArgumentException("진작 대출되어 있는 책입니다.")
        }

        val user = userRepository.findByName(request.userName) ?: fail()
        user.loanBook(book)
    }

    @Transactional
    fun returnBook(request: BookReturnRequest) {
        val user = userRepository.findByName(request.userName) ?: fail()
        user.returnBook(request.bookName)
    }

    fun countLoanedBook(): Int {
        return userLoanHistoryRepository.countByStatus(UserLoanStatus.LOANED).toInt()
    }

    fun getBookStatistics(): List<BookStatResponse> {
//        val results = mutableListOf<BookStatResponse>()
//        val books = bookRepository.findAll()
//        for (book in books) {
//            results.firstOrNull { dto -> book.type == dto.type }?.plusOne() ?: results.add(BookStatResponse(book.type, 1))
//        }
//        return results

        // groupBy 함수형 프로그래밍 이용해서 리팩토링
//        return bookRepository.findAll() // List<Book>
//            .groupBy { book -> book.type }  // Map<BookType, List<Book>>
//            .map { (type, books) -> BookStatResponse(type, books.size) }    // List<BookStatResponse>

        return bookRepository.getStatistics()
    }
}